pkg/proto/proto.go

```diff
@@ -2044,3 +2058,233 @@ type MiningLimits struct {
 	ClassicAmountOfTxsInBlock   int
 	MaxTxsSizeInBytes           int
 }
+
+func buildHeader(body []byte, messID PeerMessageID) (Header, error) {
+	var h Header
+	h.Length = maxHeaderLength + uint32(len(body)) - headerChecksumLen
+	h.Magic = headerMagic
+	h.ContentID = messID
+	h.PayloadLength = uint32(len(body))
+	dig, err := crypto.FastHash(body)
+	if err != nil {
+		return Header{}, err
+	}
+	copy(h.PayloadChecksum[:], dig[:headerChecksumLen])
+	return h, nil
+}
+
+type GetBlockSnapshotMessage struct {
+	BlockID BlockID
+}
+
+func (m *GetBlockSnapshotMessage) ReadFrom(r io.Reader) (int64, error) {
+	packet, nn, err := readPacket(r)
+	if err != nil {
+		return nn, err
+	}
+
+	return nn, m.UnmarshalBinary(packet)
+}
+
+func (m *GetBlockSnapshotMessage) WriteTo(w io.Writer) (int64, error) {
+	buf, err := m.MarshalBinary()
+	if err != nil {
+		return 0, err
+	}
+	nn, err := w.Write(buf)
+	n := int64(nn)
+	return n, err
+}
+
+func (m *GetBlockSnapshotMessage) UnmarshalBinary(data []byte) error {
+	return parsePacket(data, ContentIDGetBlockSnapshot, "GetBlockSnapshotMessage", func(payload []byte) error {
+		blockID, err := NewBlockIDFromBytes(payload)
+		if err != nil {
+			return err
+		}
+		m.BlockID = blockID
+		return nil
+	})
+}
+
+func (m *GetBlockSnapshotMessage) MarshalBinary() ([]byte, error) {
+	body := m.BlockID.Bytes()
+	h, err := buildHeader(body, ContentIDGetBlockSnapshot)
+	if err != nil {
+		return nil, err
+	}
+	hdr, err := h.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+	body = append(hdr, body...)
+	return body, nil
+}
+
+type BlockSnapshotMessage struct {
+	Bytes []byte
+}
+
+func (m *BlockSnapshotMessage) ReadFrom(r io.Reader) (int64, error) {
+	packet, nn, err := readPacket(r)
+	if err != nil {
+		return nn, err
+	}
+
+	return nn, m.UnmarshalBinary(packet)
+}
+
+func (m *BlockSnapshotMessage) WriteTo(w io.Writer) (int64, error) {
+	buf, err := m.MarshalBinary()
+	if err != nil {
+		return 0, err
+	}
+	nn, err := w.Write(buf)
+	n := int64(nn)
+	return n, err
+}
+
+func (m *BlockSnapshotMessage) UnmarshalBinary(data []byte) error {
+	if len(data) < maxHeaderLength {
```

Comment:
It's the max header length, because we have two variables `headerSizeWithPayload` == 17 and `headerSizeWithoutPayload` == 13 and `maxHeaderLength == headerSizeWithPayload`

Reply:
Got it